src/probnum/prob/_random_variable.py

```diff
@@ -0,0 +1,904 @@
+    from cached_property import cached_property
+
+import numpy as np
+import scipy.stats
+import scipy.sparse
+import scipy._lib._util
+
+from probnum import utils as _utils
+
+
+ValueType = TypeVar("ValueType")
+RandomStateType = Union[  # see scipy._lib._util.check_random_state
+    None, int, np.random.RandomState, np.random.Generator
+]
+
+
+# pylint: disable=too-many-instance-attributes,too-many-public-methods
+class RandomVariable(Generic[ValueType]):
+    """
+    Random variables are the main objects used by probabilistic numerical methods.
+
+    Every probabilistic numerical method takes a random variable encoding the prior
+    distribution as input and outputs a random variable whose distribution encodes the
+    uncertainty arising from finite computation. The generic signature of a
+    probabilistic numerical method is:
+
+    ``output_rv = probnum_method(input_rv, method_params)``
+
+    In practice, most random variables used by methods in ProbNum have Dirac or Gaussian
+    measure.
+
+    Instances of :class:`RandomVariable` can be added, multiplied, etc. with arrays and
+    linear operators. This may change their ``distribution`` and not necessarily all
+    previously available methods are retained.
+
+    Parameters
+    ----------
+    shape : tuple
+        Shape of realizations of this random variable.
+    dtype : numpy.dtype or object
+        Data type of realizations of this random variable. If ``object`` will be
+        converted to ``numpy.dtype``.
+    distribution : Distribution
+        Probability distribution of the random variable.
+
+    See Also
+    --------
+    asrandvar : Transform into a :class:`RandomVariable`.
+    Distribution : A class representing probability distributions.
+
+    Examples
+    --------
+    """
+
+    # pylint: disable=too-many-arguments,too-many-locals
+    def __init__(
+        self,
+        shape: Optional[Union[int, Tuple[int, ...]]] = None,
+        dtype: Optional[np.dtype] = None,
+        random_state: Optional[RandomStateType] = None,
+        parameters: Optional[Dict[str, Any]] = None,
+        sample: Optional[Callable[[int], ValueType]] = None,
+        in_support: Optional[Callable[[ValueType], bool]] = None,
+        pdf: Optional[Callable[[ValueType], float]] = None,
+        logpdf: Optional[Callable[[ValueType], float]] = None,
+        cdf: Optional[Callable[[ValueType], float]] = None,
+        logcdf: Optional[Callable[[ValueType], float]] = None,
+        quantile: Optional[Callable[[float], ValueType]] = None,
+        mode: Optional[Callable[[], ValueType]] = None,
+        median: Optional[Callable[[], ValueType]] = None,
+        mean: Optional[Callable[[], ValueType]] = None,
+        cov: Optional[Callable[[], ValueType]] = None,
+        var: Optional[Callable[[], ValueType]] = None,
+        std: Optional[Callable[[], ValueType]] = None,
+        entropy: Optional[Callable[[], float]] = None,
+        properties: Optional[Dict[str, Any]] = None,
```

Comment:
Let's discuss again whether it is necessary to have this `properties` arguments and if yes about its naming. It can be somewhat confusing what is acutally meant (e.g. rv properties or Python properties) when talking about `RandomVariable`'s properties.

Reply:
As discussed in person, I removed the `properties`.